File: config/src/main/java/com/typesafe/config/impl/SimpleConfigDocument.java
package com.typesafe.config.impl;

import com.typesafe.config.parser.ConfigDocument;
import com.typesafe.config.ConfigParseOptions;
import com.typesafe.config.ConfigValue;

import java.io.StringReader;
import java.util.Iterator;

final class SimpleConfigDocument implements ConfigDocument {
    private ConfigNodeRoot configNodeTree;
    private ConfigParseOptions parseOptions;

    SimpleConfigDocument(ConfigNodeRoot parsedNode, ConfigParseOptions parseOptions) {
        configNodeTree = parsedNode;
        this.parseOptions = parseOptions;
    }

    public ConfigDocument setValue(String path, String newValue) {
        SimpleConfigOrigin origin = SimpleConfigOrigin.newSimple("single value parsing");
        StringReader reader = new StringReader(newValue);
        Iterator<Token> tokens = Tokenizer.tokenize(origin, reader, parseOptions.getSyntax());
        AbstractConfigNodeValue parsedValue = ConfigDocumentParser.parseValue(tokens, origin, parseOptions);
        reader.close();

        return new SimpleConfigDocument(configNodeTree.setValue(path, parsedValue, parseOptions.getSyntax()), parseOptions);
    }

    public ConfigDocument setValue(String path, ConfigValue newValue) {
        return setValue(path, newValue.render());
    }

    public ConfigDocument removeValue(String path) {
        return new SimpleConfigDocument(configNodeTree.setValue(path, null, parseOptions.getSyntax()), parseOptions);
    }

    public boolean hasValue(String path) {
        return configNodeTree.hasValue(path);
    }

    public String render() {
        return configNodeTree.render();
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof ConfigDocument && render().equals(((ConfigDocument) other).render());
    }

    @Override
    public int hashCode() {
        return render().hashCode();
    }
}

File: config/src/main/scala/com/typesafe/config/impl/SimpleConfigDocument.scala
package com.typesafe.config.impl

import com.typesafe.config._
import com.typesafe.config.parser.ConfigDocument
import java.io.StringReader

final class SimpleConfigDocument private[impl] (
    var configNodeTree: ConfigNodeRoot,
    var parseOptions: ConfigParseOptions) extends ConfigDocument {
    override def withValueText(path: String, newValue: String): ConfigDocument = {
        if (newValue == null)
            throw new ConfigException.BugOrBroken(
                "null value for " + path + " passed to withValueText")
        val origin =
            SimpleConfigOrigin.newSimple("single value parsing")
        val reader = new StringReader(newValue)
        val tokens =
            Tokenizer.tokenize(origin, reader, parseOptions.getSyntax)
        val parsedValue =
            ConfigDocumentParser.parseValue(tokens, origin, parseOptions)
        reader.close()
        new SimpleConfigDocument(
            configNodeTree.setValue(path, parsedValue, parseOptions.getSyntax),
            parseOptions)
    }
    override def withValue(
        path: String,
        newValue: ConfigValue): ConfigDocument = {
        if (newValue == null)
            throw new ConfigException.BugOrBroken(
                "null value for " + path + " passed to withValue")
        var options = ConfigRenderOptions.defaults
        options = options.setOriginComments(false)
        withValueText(path, newValue.render(options).trim)
    }
    override def withoutPath(path: String) = new SimpleConfigDocument(
        configNodeTree.setValue(path, null, parseOptions.getSyntax),
        parseOptions)
    override def hasPath(path: String): Boolean = configNodeTree.hasValue(path)
    override def render: String = configNodeTree.render
    override def equals(other: Any): Boolean =
        other.isInstanceOf[ConfigDocument] && render == other
            .asInstanceOf[ConfigDocument]
            .render
    override def hashCode: Int = render.hashCode
}
